lib/redis_cluster/client.rb

```diff
@@ -22,29 +30,48 @@ def initialize(startup_hosts, configs = {})
     end
 
     def execute(method, args, &block)
-      ttl = Configuration::REQUEST_TTL
       asking = false
+      last_error = nil
+      retries_left = @retry_count
       try_random_node = false
 
-      while ttl > 0
-        ttl -= 1
+      # We use `>= 0` instead of `> 0` because we decrement this counter on the
+      # first run.
+      while retries_left >= 0
+        retries_left -= 1
+
         begin
           return @pool.execute(method, args, {asking: asking, random_node: try_random_node}, &block)
-        rescue Errno::ECONNREFUSED, Redis::TimeoutError, Redis::CannotConnectError, Errno::EACCES
+
+        rescue Errno::ECONNREFUSED, Redis::TimeoutError, Redis::CannotConnectError, Errno::EACCES => e
+          last_error = e
+
+          # Getting an error while executing may be an indication that we've
+          # lost the node that we were talking to and in that case it makes
+          # sense to try a different node and maybe reload our node pool (if
+          # the new node issues a `MOVE`).
           try_random_node = true
-          sleep 0.1 if ttl < Configuration::REQUEST_TTL / 2
+
         rescue => e
+          last_error = e
+
           err_code = e.to_s.split.first
           raise e unless %w(MOVED ASK).include?(err_code)
 
           if err_code == 'ASK'
             asking = true
           else
+            # `MOVED` indicates a permanent which means that our slot mappings
+            # our stale: reload them.
```

Comment:
I assume you meant:

s/permanent/permanent move/

s/our stale/are stale/

Reply:
Whoops, butchered that one. Thanks! Updated this patch and rebased everything.